File: algorithms/2_medium/S/00974_subarray_sums_divisible_by_k/table.cpp
class Solution {
public:
    static int subarraysDivByK(const vector<int>& nums, const int k) {
        assert(k >= 2 && k <= 10'000 && nums.size() <= 30'000);
        auto freqs = vector<int>(k, 0);
        int sum = 0;
        freqs[0] = 1;

        for (const int num : nums) {
            sum += num;
            freqs[((sum % k) + k) % k] += 1;
        }

        int res = 0;
        for (const int freq : freqs) {
            res += freq * (freq - 1) / 2;
        }

        return res;
    }
};

File: algorithms/2_medium/S/00974_subarray_sums_divisible_by_k/table.rs
impl Solution {
    pub fn subarrays_div_by_k(nums: Vec<i32>, k: i32) -> i32 {
        assert!(k >= 2 && k <= 10_000 && nums.len() <= 30_000);
        let mut freqs = vec![0_i32; k as usize];
        let mut sum = 0_i32;
        freqs[0] = 1;

        for num in nums {
            sum += num;
            freqs[(((sum % k) + k) % k) as usize] += 1;
        }

        freqs.into_iter().map(|x| x  * (x - 1) / 2).sum::<i32>()
    }
}
